sunpy/coordinates/utils.py

```diff
@@ -265,3 +265,72 @@ def coordinates(self, points=None):
+
+    Parameters
+    ----------
+    bottom_left : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The bottom-left coordinate of the rectangle.
+        Supports passing both ``bottom_left`` and ``top_right`` as an array of shape ``(2,)``.
+    top_right : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The top-right coordinate of the rectangle. Can only be omitted if
+        ``bottom_left`` has shape ``(2,)``.
+        If in a different frame than ``bottom_left`` and all required metadata for frame conversion is present,
+        ``top_right`` will be transformed to ``bottom_left`` frame.
+        Otherwise a ``ValueError`` will be raised.
+    width : `~astropy.units.Quantity`
+        The width of the rectangle.
+        Must be omitted if ``bottom_left`` has shape ``(2,)`` or ``top_right`` is provided.
+    height : `~astropy.units.Quantity`
+        The height of the rectangle.
+        Must be omitted if ``bottom_left`` has shape ``(2,)`` or ``top_right`` is provided.
+
+    Returns
+    -------
+    `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The bottom left coordinate of the rectangular region of interest.
+    `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The top right coordinate of the rectangular region of interest.
+    """
+    if not (hasattr(bottom_left, 'transform_to') and
+            hasattr(bottom_left, 'shape') and
+            hasattr(bottom_left, 'spherical')):
+        raise TypeError("Invalid input, bottom_left must be of type SkyCoord or BaseCoordinateFrame")
+
+    if (top_right is not None and not ((hasattr(top_right, 'transform_to') and
+                                        hasattr(top_right, 'shape') and
+                                        hasattr(top_right, 'spherical')))):
+        raise TypeError("Invalid input, top_right must be of type SkyCoord or BaseCoordinateFrame")
+
+    if bottom_left.shape == (2,) and any((x is not None for x in (width, height, top_right))):
+        raise ValueError("Invalid input, if bottom_left.shape == (2,) "
+                         "other parameters should not be passed.")
+
+    if all(x is not None for x in (width, height, top_right)):
+        raise ValueError("Invalid input, width, height and top_right "
+                         "parameters should not be passed simultaneously.")
+
+    if top_right is None and bottom_left.shape != (2,) and not all((width, height)):
+        raise ValueError("Invalid input, either bottom_left and top_right "
+                         "or bottom_left and height and width should be provided.")
+
+    if bottom_left.shape == (2,):
+        top_right = bottom_left[1]
+        bottom_left = bottom_left[0]
+
+    elif top_right is not None:
+        top_right = top_right.transform_to(bottom_left)
+
+    else:
+        top_right = SkyCoord(bottom_left.spherical.lon + width,
```

Comment:
Add a comment here about type casting, as discussed in a couple of threads.

(new comment to track it)